Add Transfer Entry for Devicetree Overlay

raymo200915 opened the pull request:
Add XFERLIST_DT_OVERLAY to represent Devicetree Overlay Notes which are being used in U-Boot and kernel to update the live devicetree.

raymo200915 committed:
Add Transfer Entry for Devicetree Overlay

Add XFERLIST_DT_OVERLAY to represent Devicetree Overlay Notes which
are being used in U-Boot and kernel to update the live devicetree.

Signed-off-by: Raymond Mao <raymond.mao@linaro.org>

sjg20 commented on source/transfer_list.rst:
```diff
@@ -837,6 +837,47 @@ and reserved for use as a TPM Command Response Buffer interface.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Devicetree Overlay is defined in [DTO]_. The Devicetree Overlay TE contains
+the Device Overlay Note in the data section.
```
Does this refer to https://docs.kernel.org/devicetree/dynamic-resolution-notes.html ? If so, can you please add a link to it?

Is this information available in [DT]. I couldn't find it but I might have missed something.

raymo200915 replied:
Reference on [DTO] is fair enough, since it refers to https://docs.kernel.org/devicetree/dynamic-resolution-notes.html.
I cannot find this information from [DT] too.

sjg20 commented on source/transfer_list.rst:
```diff
@@ -837,6 +837,47 @@ and reserved for use as a TPM Command Response Buffer interface.
+The Devicetree Overlay is defined in [DTO]_. The Devicetree Overlay TE contains
+the Device Overlay Note in the data section.
+The intent of the Devicetree Overlay entry is to modify the live devicetree in
+U-Boot or kernel.
```
Can you please add a few more details about the lifefcycle here?
- Where does the overlay come from? 
- Where does the base devicetree come from?
- Why the overlay is needed in addition to the base devicetree (different types of hardware / variants?)
- Are the devicetrees signed? How does security work with this?
- When should the overlay be applied? In U-Boot, or Linux, or something else?
- Give an example of projects in the boot chain and how this would be used

raymo200915 commented on source/transfer_list.rst:
```diff
@@ -837,6 +837,47 @@ and reserved for use as a TPM Command Response Buffer interface.
+   * - dto
+     - data_size
+     - hdr_size
+     - The dto field contains the DT overlay.
```
Yes, an index field (or subtype) will be added into the TE header (or the unused bits of tag ID) as a general solution to all multiple TEs. @manish-pandey-arm has proposed the solution.
So we don't need to add this specially for DT or DTO.

danh-arm replied:
I'm not sure that we are going to extend the TE header at this stage so we may need to add a subtype to specific TEs like this. I believe there's a plan to schedule a call on this subject with the other maintainers. So let's keep this on hold until then.

raymo200915 replied:
> I'm not sure that we are going to extend the TE header at this stage so we may need to add a subtype to specific TEs like this. I believe there's a plan to schedule a call on this subject with the other maintainers. So let's keep this on hold until then.

FMO, having it in the header is more convenient than in the payload. Yes, this PR is not urgent to be merged. I just want to reserve the placeholder of ID to represent DTO as I need the ID for code implementations.

danh-arm replied:
I think using a compatible node in the DTBO sounds like good idea. However, even then there may be multiple overlays targeting the same compatible node, right? So a sub-type still might be needed here.

danh-arm commented on source/transfer_list.rst:
```diff
@@ -837,6 +837,47 @@ and reserved for use as a TPM Command Response Buffer interface.
+DT overlay entry layout (XFERLIST_DT_OVERLAY)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The Devicetree Overlay is defined in [DTO]_. The Devicetree Overlay TE contains
```
The current spec doesn't have provision for multiple TEs of the same tag id:
> It is strongly recommended that entries of a given type are unique in the TL. If firmware designs require multiple TE instances of a given type, then that TE type definition should provide sufficient information for the TE consumer to disambiguate between all TE instances. That information can be, for example, a sub-type field or contained within a self-describing data blob.

However, I guess for XFERLIST_FDT, the "model = " node might be enough to distinguish between instances.

I'll assume it is important for you to distinguish between base and overlay DTs (though you didn't explicitly say that).

jwerner-chromium replied:
> So can this be renamed to something like XFERLIST_FDT_AUXILIARY to cater for both cases (multiple base DTs and multiple overlays)?

Honestly, I don't think it's a good idea to try to make tags more flexible so they can be used for many different cases... on the contrary, I think it is important to make sure each tag ID clearly identifies one single application. I agree with Simon above that this tag seems underspecified as it is already — it just says that there's a device tree overlay in this TE, but not what that overlay is for or what the receiving firmware is supposed to do with it.

If the interoperability goal is that two boot stages which haven't intentionally been co-designed can hand off data to each other and use what they were given appropriately, then the spec needs to be very precise about what each piece of data means and in what ways it may (and may not) be used. A device tree that's supposed to be used by a kernel-loading boot stage to pass to the kernel is not the same as a device tree that's supposed to be used by the boot stage itself to enumerate devices. An overlay that's supposed to be merged into a device tree that the following boot stage got from somewhere else (e.g. hardcoded in the image) is probably not the same as an overlay that's supposed to be merged into a device tree that was also passed in the TL by the same stage.

Trying to make a single tag that could be valid for all these cases probably just means that boot stages will always ignore it unless they have been specifically hardcoded to interpret it as one specific of these uses, and that kinda defeats the purpose of self-describing data (because then an important piece of the information, how exactly to use it, is still passed out of band through compile-time configuration). If the tags were more precise (e.g. XFERLIST_DTBO_MERGE_INTO_KERNEL), the chance that things would interoperate correctly without both sides being specifically hardcoded on what to expect from the other would be higher.

If the idea is that some description inside the DTBO should resolve this ambiguity, then I think this spec should exactly describe the fields that should be used for that (so that all implementations can be sure to agree on the same thing).

raymo200915 replied:
> If the idea is that some description inside the DTBO should resolve this ambiguity, then I think this spec should exactly describe the fields that should be used for that (so that all implementations can be sure to agree on the same thing)

The idea is that all self-described TE payloads (not limited to DT/DTO) should solve this ambiguity.
For the non-self-described ones, I think we cannot do anything now unless a 'subtype' is introduced.

raymo200915 replied:
> > The idea is that all self-described TE payloads (not limited to DT/DTO) should solve this ambiguity.
> 
> Yes, but how? You can't just say "this is a DTBO and it describes what it is being used for". DTBOs don't have any universally accepted fields that clarify whether they're e.g. supposed to be applied to a Linux kernel or a firmware stage like U-Boot. If you want there to be something like a property at the root node of the DTBO that makes this distinction, you need to write in this spec how that property is called, what its possible values are and what they mean. Otherwise, how should an implementation that just reads the spec and wants to implement support for this tag know what to do with it?
> 
> > For the non-self-described ones, I think we cannot do anything now unless a 'subtype' is introduced.
> 
> Well, the solution would be to make tags more precise until they only mean a single thing. For example, the XFERLIST_EVLOG tag definition is pretty precise: it says that it contains a log of all measurements extended into the TPM, and the log format and extension mechanism are defined by a different well-known spec. The ambiguity of whether it contains measurements already extended or measurements cached to be extended in the future is resolved by a flags field. So any firmware stage that does TPM measurements and recognizes this tag can know how to use it and add its own measurements there. There's only one way to make TPM measurements so different stages by different implementors can't really disagree on how to do that.
> 
> The tag you're adding here is different. It just says "there's a device tree overlay". It doesn't say what that overlay contains or how you're supposed to use it, and there are many different ways in which device tree overlays can be used. I think the discussion of how to best encode the necessary additional information to clarify how exactly a TE received with this tag can be used (e.g. with extra fields in the TE, or by splitting different cases onto different tag IDs) should be part of this design process.

Actually the tag I am adding now is for being used in U-Boot. Since U-Boot ends up the lifecycle of TL and will merge all overlays handed from previous stages before launching kernel. This is a straightforward usecase and similar with the `XFERLIST_DTBO_MERGE_INTO_KERNEL` you proposed.
If adding a general DTO TE has concerns on how to precisely describe its usage, I can rephase and limit it to "DTOs to be merged for kernel". Is that good?

apalos replied:
> > The idea is that all self-described TE payloads (not limited to DT/DTO) should solve this ambiguity.
> 
> Yes, but how? You can't just say "this is a DTBO and it describes what it is being used for". DTBOs don't have any universally accepted fields that clarify whether they're e.g. supposed to be applied to a Linux kernel or a firmware stage like U-Boot. 

What's the concern here though? A DT is described by a spec and no entry should be invalid. IOW you should be able to apply overlays regardless of the consumer. Are you worried about size limitations?

>If you want there to be something like a property at the root node of the DTBO that makes this distinction, you need to write in this spec how that property is called, what its possible values are and what they mean. Otherwise, how should an implementation that just reads the spec and wants to implement support for this tag know what to do with it?
> 
> > For the non-self-described ones, I think we cannot do anything now unless a 'subtype' is introduced.
> 
> Well, the solution would be to make tags more precise until they only mean a single thing. For example, the XFERLIST_EVLOG tag definition is pretty precise: it says that it contains a log of all measurements extended into the TPM, and the log format and extension mechanism are defined by a different well-known spec. The ambiguity of whether it contains measurements already extended or measurements cached to be extended in the future is resolved by a flags field. So any firmware stage that does TPM measurements and recognizes this tag can know how to use it and add its own measurements there. There's only one way to make TPM measurements so different stages by different implementors can't really disagree on how to do that.

I don't know all the use cases and people doing products might have a better insight, but don't we have cases that DTBOs have to be applied unconditionally? If so, we can consider this our catch-all entry.

> 
> The tag you're adding here is different. It just says "there's a device tree overlay". It doesn't say what that overlay contains or how you're supposed to use it, and there are many different ways in which device tree overlays can be used. I think the discussion of how to best encode the necessary additional information to clarify how exactly a TE received with this tag can be used (e.g. with extra fields in the TE, or by splitting different cases onto different tag IDs) should be part of this design process.

Does anyone have use cases where the DTBO has to be purged after being consumed by a boot stage?

raymo200915 replied:
> > If adding a general DTO TE has concerns on how to precisely describe its usage, I can rephase and limit it to "DTOs to be merged for kernel". Is that good?
> 
> Yeah, I think that would make a lot more sense. Does the U-Boot in your case also apply this overlay to the DTB it uses for its own device enumeration, or only to the one it passes to the kernel? That's part of the distinction that should also be written down.
> 
I am going to rename this as `XFERLIST_DT_OVERLAY_KERNEL_LOADER`, that is for a kernel loader (not limited to U-Boot) to apply into its own device tree and finally pass to the kernel - This is the only usecase we have at the moment.
When we decide to choose this approach, we won't have a tag for general purpose DTO leveraging the self-describability in the future but each of the DTO has to define its own tag specifying the purpose, just want to make sure if all of you agree with. 
@apalos @jwerner-chromium @sjg20 @danh-arm @manish-pandey-arm @harrisonmutai-arm

danh-arm replied:
I agree with Julius's general point above that a given tagid should refer to the specific purpose of the DT, so in this case an overlay to the kernel DT. However, I'm not seeing the use-case for distinguishing between a DT used by a firmware for its own device enumeration and the DT passed to the kernel. In all the platforms I've seen, it's the same DT that gets used/modified through the various firmwares until it reaches the kernel. If there is such a use-case, then we should add another base DT tagid for the former case.

So I'd suggest naming this XFERLIST_HW_FDT_OVERLAY or XFERLIST_KERNEL_FDT_OVERLAY. If we go with that, then it might be clearer if we renamed the existing XFERLIST_FDT to match (though that can be a separate patch).

danh-arm replied:
> We do have usecase that BL2 generates an overlay for applying in BL33, but it is not expected to be applied in BL31/32.

Interesting, but are you just saying the overlay is probably not relevant for BL31/32 or the overlay *must* not be applied to BL31/32? For the former case, I don't think the spec needs to make a distinction between BL33 and earlier stages - it can be IMP DEF whether a stage applies the overlay. For the latter case, I agree a distinction is necessary.

> Maybe naming it to XFERLIST_FDT_OVERLAY_BL33 is more helpful.

BL33 is TF-A terminology. If a distinction is necessary, perhaps XFERLIST_FDT_NORMAL_WORLD_OVERLAY?

sjg20 commented:
I just wanted to mention here (as I did in the meeting) the FIT spec, which could be a good format for the information needed here.

https://fitspec.osfw.foundation/#document-chapter2-source-file-format

raymo200915 commented:
The patch is amended with below items per our discussions on Jun 19 sync-up:
1. added description on DTO applying based on the matching top-level compatible string.
2. added subtype.
3. minor misc fixes.

A patch was posted to linux-doc to update the top-level compatible in DTSO at:
https://lore.kernel.org/linux-doc/20250624181320.2810521-1-raymond.mao@linaro.org/T/#u

sjg20 reviewed (commented):
I'm not sure if this proposal is still live. But I think we should adopt FIT as the standard way of shipping a collection of device trees/overlays, rather than implementing it piecemeal.

There was also a proposal here:

https://lore.kernel.org/devicetree-spec/CAD=FV=Vrit=N6L30dA3X6uV95e5E1bVoLph55_=ihqSTBy52FQ@mail.gmail.com/T/#t

michalsimek commented:
I have read this thread and it looks like that it is stuck. What would be the way out? @manish-pandey-arm , @danh-arm  and @apalos . We have TF-A patches ready for run time memory reservation injection done by TF-A and also by TEE which we want to send to U-Boot to put it together. The reason is that we a lot of misconfiguration and using DT overlays will make sure that NS sw has proper memory reservation described. 
We are not generating any compatible string for overlay itself but that's likely a logic in our case for U-Boot to deal with it. From TL perspective we just need to mark that this is DT overlay. 
Question if make sense to apply it should be deferred to SW which should consume it. 
That's why I would like to move this topic to be able to upstream code to TF-A and OP-TEE.

sjg20 commented:
> I have read this thread and it looks like that it is stuck. What would be the way out? @manish-pandey-arm , @danh-arm and @apalos . We have TF-A patches ready for run time memory reservation injection done by TF-A and also by TEE which we want to send to U-Boot to put it together. The reason is that we a lot of misconfiguration and using DT overlays will make sure that NS sw has proper memory reservation described. We are not generating any compatible string for overlay itself but that's likely a logic in our case for U-Boot to deal with it. From TL perspective we just need to mark that this is DT overlay. Question if make sense to apply it should be deferred to SW which should consume it. That's why I would like to move this topic to be able to upstream code to TF-A and OP-TEE.

Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?

michalsimek commented:
> > I have read this thread and it looks like that it is stuck. What would be the way out? @manish-pandey-arm , @danh-arm and @apalos . We have TF-A patches ready for run time memory reservation injection done by TF-A and also by TEE which we want to send to U-Boot to put it together. The reason is that we a lot of misconfiguration and using DT overlays will make sure that NS sw has proper memory reservation described. We are not generating any compatible string for overlay itself but that's likely a logic in our case for U-Boot to deal with it. From TL perspective we just need to mark that this is DT overlay. Question if make sense to apply it should be deferred to SW which should consume it. That's why I would like to move this topic to be able to upstream code to TF-A and OP-TEE.
> 
> Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?

What we have is reserved memory node where TF-A, OP-TEE just mark memory regions as reserved based on their own configuration. PSCI node should be also there. CPU topology is something what I am thinking about because we want to support different CPU configurations (some cpu disabled via efuse, lockstep configurations, etc) via same firmware.

IIRC U-Boot is using own IDs but didn't see that in the spec. I am fine if answer is simple that we should use own SOC id for this feature and maintain ID list our self. And we will just extend ID which are available in U-Boot's include/bloblist.h

        /* Other standard area to allocate blobs */
        BLOBLISTT_AREA_OTHER = 0x200,
        BLOBLISTT_INTEL_VBT = 0x200,    /* Intel Video-BIOS table */
        BLOBLISTT_SMBIOS_TABLES = 0x201, /* SMBIOS tables for x86 */
        BLOBLISTT_VBOOT_CTX = 0x202,    /* Chromium OS verified boot context */

sjg20 commented:
> > Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?
> 
> What we have is reserved memory node where TF-A, OP-TEE just mark memory regions as reserved based on their own configuration. PSCI node should be also there. CPU topology is something what I am thinking about because we want to support different CPU configurations (some cpu disabled via efuse, lockstep configurations, etc) via same firmware.
> 
> IIRC U-Boot is using own IDs but didn't see that in the spec. I am fine if answer is simple that we should use own SOC id for this feature and maintain ID list our self. And we will just extend ID which are available in U-Boot's include/bloblist.h
> 
> ```
>     /* Other standard area to allocate blobs */
>     BLOBLISTT_AREA_OTHER = 0x200,
>     BLOBLISTT_INTEL_VBT = 0x200,    /* Intel Video-BIOS table */
>     BLOBLISTT_SMBIOS_TABLES = 0x201, /* SMBIOS tables for x86 */
>     BLOBLISTT_VBOOT_CTX = 0x202,    /* Chromium OS verified boot context */
> ```

How about we start with the reserved-memory feature and document that this is available via this tag? My concern is that a a general DT-overlay feature would be quite open-ended. But if we can scope it to just reserved-memory for now, then go from there, it should be fine.

michalsimek commented:
> > > Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?
> > 
> > 
> > What we have is reserved memory node where TF-A, OP-TEE just mark memory regions as reserved based on their own configuration. PSCI node should be also there. CPU topology is something what I am thinking about because we want to support different CPU configurations (some cpu disabled via efuse, lockstep configurations, etc) via same firmware.
> > IIRC U-Boot is using own IDs but didn't see that in the spec. I am fine if answer is simple that we should use own SOC id for this feature and maintain ID list our self. And we will just extend ID which are available in U-Boot's include/bloblist.h
> > ```
> >     /* Other standard area to allocate blobs */
> >     BLOBLISTT_AREA_OTHER = 0x200,
> >     BLOBLISTT_INTEL_VBT = 0x200,    /* Intel Video-BIOS table */
> >     BLOBLISTT_SMBIOS_TABLES = 0x201, /* SMBIOS tables for x86 */
> >     BLOBLISTT_VBOOT_CTX = 0x202,    /* Chromium OS verified boot context */
> > ```
> 
> How about we start with the reserved-memory feature and document that this is available via this tag? My concern is that a a general DT-overlay feature would be quite open-ended. But if we can scope it to just reserved-memory for now, then go from there, it should be fine.

Who will be limit this? I mean spec will say here is tag for dt overlay for reserved memory but who will limit it? Pretty much if this is reserved memory node and other things when this get to U-Boot, U-Boot will just apply them. Or do you want in u-boot to start to check that reserved memory node overlay is the only component there?

sjg20 commented:
> > > > Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?
> > > 
> > > 
> > > What we have is reserved memory node where TF-A, OP-TEE just mark memory regions as reserved based on their own configuration. PSCI node should be also there. CPU topology is something what I am thinking about because we want to support different CPU configurations (some cpu disabled via efuse, lockstep configurations, etc) via same firmware.
> > > IIRC U-Boot is using own IDs but didn't see that in the spec. I am fine if answer is simple that we should use own SOC id for this feature and maintain ID list our self. And we will just extend ID which are available in U-Boot's include/bloblist.h
> > > ```
> > >     /* Other standard area to allocate blobs */
> > >     BLOBLISTT_AREA_OTHER = 0x200,
> > >     BLOBLISTT_INTEL_VBT = 0x200,    /* Intel Video-BIOS table */
> > >     BLOBLISTT_SMBIOS_TABLES = 0x201, /* SMBIOS tables for x86 */
> > >     BLOBLISTT_VBOOT_CTX = 0x202,    /* Chromium OS verified boot context */
> > > ```
> > 
> > 
> > How about we start with the reserved-memory feature and document that this is available via this tag? My concern is that a a general DT-overlay feature would be quite open-ended. But if we can scope it to just reserved-memory for now, then go from there, it should be fine.
> 
> Who will be limit this? I mean spec will say here is tag for dt overlay for reserved memory but who will limit it? Pretty much if this is reserved memory node and other things when this get to U-Boot, U-Boot will just apply them. Or do you want in u-boot to start to check that reserved memory node overlay is the only component there?

One way is to say that if other things are added the behaviour is undefined and U-Boot may ignore the entire overlay, for example. Another way is to implement such a check in U-Boot.

I wonder if we should specify the reserved memory in a different way (e.g. a separate tag) rather than relying on something as general as a DT?

michalsimek commented:
> > > > > Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?
> > > > 
> > > > 
> > > > What we have is reserved memory node where TF-A, OP-TEE just mark memory regions as reserved based on their own configuration. PSCI node should be also there. CPU topology is something what I am thinking about because we want to support different CPU configurations (some cpu disabled via efuse, lockstep configurations, etc) via same firmware.
> > > > IIRC U-Boot is using own IDs but didn't see that in the spec. I am fine if answer is simple that we should use own SOC id for this feature and maintain ID list our self. And we will just extend ID which are available in U-Boot's include/bloblist.h
> > > > ```
> > > >     /* Other standard area to allocate blobs */
> > > >     BLOBLISTT_AREA_OTHER = 0x200,
> > > >     BLOBLISTT_INTEL_VBT = 0x200,    /* Intel Video-BIOS table */
> > > >     BLOBLISTT_SMBIOS_TABLES = 0x201, /* SMBIOS tables for x86 */
> > > >     BLOBLISTT_VBOOT_CTX = 0x202,    /* Chromium OS verified boot context */
> > > > ```
> > > 
> > > 
> > > How about we start with the reserved-memory feature and document that this is available via this tag? My concern is that a a general DT-overlay feature would be quite open-ended. But if we can scope it to just reserved-memory for now, then go from there, it should be fine.
> > 
> > 
> > Who will be limit this? I mean spec will say here is tag for dt overlay for reserved memory but who will limit it? Pretty much if this is reserved memory node and other things when this get to U-Boot, U-Boot will just apply them. Or do you want in u-boot to start to check that reserved memory node overlay is the only component there?
> 
> One way is to say that if other things are added the behaviour is undefined and U-Boot may ignore the entire overlay, for example. Another way is to implement such a check in U-Boot.
> 
> I wonder if we should specify the reserved memory in a different way (e.g. a separate tag) rather than relying on something as general as a DT?

And what's the concern with generic overlay? I would like to understand your concern better.

sjg20 commented:
> > > > > > Just to be clear, can you give an example of a DT overlay this would support? It seems as if it might not be a real overlay (capable of making arbitrary DT changes) but just a node (and compatible string) containing some project-specific information. Is that right?
> > > > > 
> > > > > 
> > > > > What we have is reserved memory node where TF-A, OP-TEE just mark memory regions as reserved based on their own configuration. PSCI node should be also there. CPU topology is something what I am thinking about because we want to support different CPU configurations (some cpu disabled via efuse, lockstep configurations, etc) via same firmware.
> > > > > IIRC U-Boot is using own IDs but didn't see that in the spec. I am fine if answer is simple that we should use own SOC id for this feature and maintain ID list our self. And we will just extend ID which are available in U-Boot's include/bloblist.h
> > > > > ```
> > > > >     /* Other standard area to allocate blobs */
> > > > >     BLOBLISTT_AREA_OTHER = 0x200,
> > > > >     BLOBLISTT_INTEL_VBT = 0x200,    /* Intel Video-BIOS table */
> > > > >     BLOBLISTT_SMBIOS_TABLES = 0x201, /* SMBIOS tables for x86 */
> > > > >     BLOBLISTT_VBOOT_CTX = 0x202,    /* Chromium OS verified boot context */
> > > > > ```
> > > > 
> > > > 
> > > > How about we start with the reserved-memory feature and document that this is available via this tag? My concern is that a a general DT-overlay feature would be quite open-ended. But if we can scope it to just reserved-memory for now, then go from there, it should be fine.
> > > 
> > > 
> > > Who will be limit this? I mean spec will say here is tag for dt overlay for reserved memory but who will limit it? Pretty much if this is reserved memory node and other things when this get to U-Boot, U-Boot will just apply them. Or do you want in u-boot to start to check that reserved memory node overlay is the only component there?
> > 
> > 
> > One way is to say that if other things are added the behaviour is undefined and U-Boot may ignore the entire overlay, for example. Another way is to implement such a check in U-Boot.
> > I wonder if we should specify the reserved memory in a different way (e.g. a separate tag) rather than relying on something as general as a DT?
> 
> And what's the concern with generic overlay? I would like to understand your concern better.

It is very open-ended. The firmware could change almost anything such that the OS won't boot, etc. I would prefer to have a clear interface for the specific use cases rather than something which seems like a back door to me :-)

michalsimek commented:
> > > 
> > > One way is to say that if other things are added the behaviour is undefined and U-Boot may ignore the entire overlay, for example. Another way is to implement such a check in U-Boot.
> > > I wonder if we should specify the reserved memory in a different way (e.g. a separate tag) rather than relying on something as general as a DT?
> > 
> > 
> > And what's the concern with generic overlay? I would like to understand your concern better.
> 
> It is very open-ended. The firmware could change almost anything such that the OS won't boot, etc. I would prefer to have a clear interface for the specific use cases rather than something which seems like a back door to me :-)

What about to change subtype to be used as category which DT overlay targets. It means subtype 0 - unlimited, 1 - reserved-memory, subtype 2 - cpu topology, 3 - psci, etc . Then bootloader can configure ourself which DT overlays are acceptable and which one not. It means you can configure your system in a way completely ignore DT overlays or just allowing only specific parts.

danh-arm reviewed (approved):
I'm fine with proceeding with this PR as-is. I don't think it's necessary to specify particular DTBO use-cases, just how the DTBO should be applied. Firmware is already responsible for passing the kernel DT to the next stage. Passing one or more DTBOs as well doesn't materially increase the risk of the kernel eventually receiving a bad DT. The DT and all the DTBOs should be part of the authenticated boot chain.

sjg20 commented:
> > > > One way is to say that if other things are added the behaviour is undefined and U-Boot may ignore the entire overlay, for example. Another way is to implement such a check in U-Boot.
> > > > I wonder if we should specify the reserved memory in a different way (e.g. a separate tag) rather than relying on something as general as a DT?
> > > 
> > > 
> > > And what's the concern with generic overlay? I would like to understand your concern better.
> > 
> > 
> > It is very open-ended. The firmware could change almost anything such that the OS won't boot, etc. I would prefer to have a clear interface for the specific use cases rather than something which seems like a back door to me :-)
> 
> What about to change subtype to be used as category which DT overlay targets. It means subtype 0 - unlimited, 1 - reserved-memory, subtype 2 - cpu topology, 3 - psci, etc . Then bootloader can configure ourself which DT overlays are acceptable and which one not. It means you can configure your system in a way completely ignore DT overlays or just allowing only specific parts.

Yes that would be a logical approach, but it is also quite strange, since we are basically adding metadata for a DT (which of course may in fact not be correct)...the bootloader could examine the DT for itself. I was more thinking of describing better (in this spec) what the feature is actually for.

Also Dan seems comfortable with just leaving this as it is.

Can you tell me how this overlay works when we load a FIT supporting (say) 10 boards. I presume the overlay would be generic to all of  them, so the selection happens in U-Boot proper. Or does the earlier phase know the board that will (later) be chosen, so will craft the overlay to target the board that it knows will be selected in the end?

michalsimek commented:
> > What about to change subtype to be used as category which DT overlay targets. It means subtype 0 - unlimited, 1 - reserved-memory, subtype 2 - cpu topology, 3 - psci, etc . Then bootloader can configure ourself which DT overlays are acceptable and which one not. It means you can configure your system in a way completely ignore DT overlays or just allowing only specific parts.
> 
> Yes that would be a logical approach, but it is also quite strange, since we are basically adding metadata for a DT (which of course may in fact not be correct)...the bootloader could examine the DT for itself. I was more thinking of describing better (in this spec) what the feature is actually for.
> 
> Also Dan seems comfortable with just leaving this as it is.
> 
> Can you tell me how this overlay works when we load a FIT supporting (say) 10 boards. I presume the overlay would be generic to all of them, so the selection happens in U-Boot proper. Or does the earlier phase know the board that will (later) be chosen, so will craft the overlay to target the board that it knows will be selected in the end?

IIRC FIT is not supported in the TL spec yet, but we are already using FIT images with the U-Boot DTB reselection feature (enabled by default) on ZynqMP Kria SOMs. Instead of a single DTB, a FIT image is passed and U-Boot, based on SOM and CC (Carrier Card) detection, selects the appropriate DT from the FIT image.

With DT overlay support in TL, I can imagine a future setup where something like XFERLIST_FIT contains a FIT image covering multiple SOM+CC combinations, and TF-A/TEE inject DT overlays (with memory reservations today) into the TL. When BL33 (U-Boot) receives this list, it will first perform SOM/CC detection and select the DT for that configuration. Based on BL33 configuration, overlays can be applied or rejected. The TL simply carries them.

What exactly goes into an overlay is up to each vendor or product to decide. Overlays can be generic and applied to all board combinations (such as memory reservations), or they can be board/SOC specific based on autodetection in TF-A/TEE. Both approaches may or may not include a compatible string to filter application by BL33.

sjg20 commented:
> > > What about to change subtype to be used as category which DT overlay targets. It means subtype 0 - unlimited, 1 - reserved-memory, subtype 2 - cpu topology, 3 - psci, etc . Then bootloader can configure ourself which DT overlays are acceptable and which one not. It means you can configure your system in a way completely ignore DT overlays or just allowing only specific parts.
> > 
> > 
> > Yes that would be a logical approach, but it is also quite strange, since we are basically adding metadata for a DT (which of course may in fact not be correct)...the bootloader could examine the DT for itself. I was more thinking of describing better (in this spec) what the feature is actually for.
> > Also Dan seems comfortable with just leaving this as it is.
> > Can you tell me how this overlay works when we load a FIT supporting (say) 10 boards. I presume the overlay would be generic to all of them, so the selection happens in U-Boot proper. Or does the earlier phase know the board that will (later) be chosen, so will craft the overlay to target the board that it knows will be selected in the end?
> 
> IIRC FIT is not supported in the TL spec yet, but we are already using FIT images with the U-Boot DTB reselection feature (enabled by default) on ZynqMP Kria SOMs. Instead of a single DTB, a FIT image is passed and U-Boot, based on SOM and CC (Carrier Card) detection, selects the appropriate DT from the FIT image.
> 
> With DT overlay support in TL, I can imagine a future setup where something like XFERLIST_FIT contains a FIT image covering multiple SOM+CC combinations, and TF-A/TEE inject DT overlays (with memory reservations today) into the TL. When BL33 (U-Boot) receives this list, it will first perform SOM/CC detection and select the DT for that configuration. Based on BL33 configuration, overlays can be applied or rejected. The TL simply carries them.
> 
> What exactly goes into an overlay is up to each vendor or product to decide. Overlays can be generic and applied to all board combinations (such as memory reservations), or they can be board/SOC specific based on autodetection in TF-A/TEE. Both approaches may or may not include a compatible string to filter application by BL33.

Thanks for the details on this. IMO it makes most sense to do the board detection very early, so that the compatible stringlist is known and can be used for overlay selection / passing through TL. If that is not the case, i.e. the compatible string is only determined in U-Boot itself, then a FIT should be transferrerd from early stages (not yet defined in TL) to contain a set of overlays for U-Boot to use.

The first approach matches most cleanly with this PR, so it seems fine to me.
